clarifai/client/input.py

```diff
@@ -0,0 +1,508 @@
+
+    def delete_inputs(self, inputs: List[Input]) -> None:
+        """Delete list of input objects from the app.
+        Args:
+            input_ids (Input): List of input objects to delete.
+        Example:
+            >>> from clarifai.client.user import User
+            >>> input_obj = User(user_id="user_id").app(app_id="app_id").inputs()
+            >>> input_obj.delete_inputs(input_obj.list_inputs())
+        """
+        if not isinstance(inputs, list):
+            raise Exception("input_ids must be a list of input ids")
+        inputs_ids = [input.id for input in inputs]
+        request = service_pb2.DeleteInputsRequest(user_app_id=self.user_app_id, ids=inputs_ids)
+        response = self._grpc_request(self.STUB.DeleteInputs, request)
+        if response.status.code != status_code_pb2.SUCCESS:
+            raise Exception(response.status)
+        self.logger.info("\nInputs Deleted\n%s", response.status)
+
+    def list_inputs(self) -> List[Input]:  # TODO: update lister
+        """Lists all the inputs for the app.
+        Returns:
+            list of Input: A list of Input objects for the app.
+        Example:
+            >>> from clarifai.client.user import User
+            >>> input_obj = User(user_id="user_id").app(app_id="app_id").inputs()
+            >>> input_obj.list_inputs()
+        """
+        request_data = dict(user_app_id=self.user_app_id, per_page=self.default_page_size)
+        all_inputs_info = list(
+            self.list_all_pages_generator(self.STUB.ListInputs, service_pb2.ListInputsRequest,
+                                        request_data))
+        for input_info in all_inputs_info:
+            input_info['id'] = input_info.pop('input_id')
+        return [resources_pb2.Input(**input_info) for input_info in all_inputs_info]
+
+    def _bulk_upload(self, inputs: List[Input], chunk_size: int = 128) -> None:
+        """Uploads process for large number of inputs .
+        Args:
+        inputs (List[Input]): input protos
+        chunk_size (int): chunk size for each request
+        """
+        num_workers: int = min(10, cpu_count())  # limit max workers to 10
+        chunk_size = min(128, chunk_size)  # limit max protos in a req
+        chunked_inputs = Chunker(inputs, chunk_size).chunk()
+        with ThreadPoolExecutor(max_workers=num_workers) as executor:
+            with tqdm(total=len(chunked_inputs), desc='Uploading inputs') as progress:
+                # Submit all jobs to the executor and store the returned futures
+                futures = [
+                    executor.submit(self._upload_batch, batch_input_ids)
+                    for batch_input_ids in chunked_inputs
+                ]
+
+                for job in as_completed(futures):
+                    retry_input_proto = job.result()
+                    self._retry_uploads(retry_input_proto)
+                    progress.update()
+
+    def _wait_for_inputs(self, input_job_id: str) -> bool:
```

Comment:
Some of the code here looks repeated from the datasets api utils. Looks like you have to figure out how to better organize the shared functions into a utils script or similar so that they are reused across modules that require them without duplication

Reply:
Exactly💯  Thats the goal.